In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('Anime_data.csv')
data.head()

Anime_id                            Title  \
0         1                     Cowboy Bebop   
1         5  Cowboy Bebop: Tengoku no Tobira   
2         6                           Trigun   
3         7               Witch Hunter Robin   
4         8                   Bouken Ou Beet   

                                               Genre  \
0  ['Action', 'Adventure', 'Comedy', 'Drama', 'Sc...   
1  ['Action', 'Space', 'Drama', 'Mystery', 'Sci-Fi']   
2  ['Action', 'Sci-Fi', 'Adventure', 'Comedy', 'D...   
3  ['Action', 'Magic', 'Police', 'Supernatural', ...   
4  ['Adventure', 'Fantasy', 'Shounen', 'Supernatu...   

                                            Synopsis   Type  \
0  In the year 2071, humanity has colonized sever...     TV   
1  Another day, another bounty—such is the life o...  Movie   
2  Vash the Stampede is the man with a $$60,000,0...     TV   
3  Witches are individuals with special powers li...     TV   
4  It is the dark century and the people are suff...     TV   

                       Producer              Studio  Rating  ScoredBy  \
0             ['Bandai Visual']         ['Sunrise']    8.81  363889.0   
1  ['Sunrise', 'Bandai Visual']           ['Bones']    8.41  111187.0   
2      ['Victor Entertainment']        ['Madhouse']    8.31  197451.0   
3             ['Bandai Visual']         ['Sunrise']    7.34   31875.0   
4                           NaN  ['Toei Animation']    7.04    4757.0   

   Popularity   Members  Episodes    Source                         Aired  \
0        39.0  704490.0      26.0  Original   Apr 3, 1998 to Apr 24, 1999   
1       475.0  179899.0       1.0  Original                   Sep 1, 2001   
2       158.0  372709.0      26.0     Manga   Apr 1, 1998 to Sep 30, 1998   
3      1278.0   74889.0      26.0  Original   Jul 2, 2002 to Dec 24, 2002   
4      3968.0   11247.0      52.0     Manga  Sep 30, 2004 to Sep 29, 2005   

                                                Link  
0       https://myanimelist.net/anime/1/Cowboy_Bebop  
1  https://myanimelist.net/anime/5/Cowboy_Bebop__...  
2             https://myanimelist.net/anime/6/Trigun  
3  https://myanimelist.net/anime/7/Witch_Hunter_R...  
4     https://myanimelist.net/anime/8/Bouken_Ou_Beet

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17002 entries, 0 to 17001
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Anime_id    17002 non-null  int64  
 1   Title       17002 non-null  object 
 2   Genre       14990 non-null  object 
 3   Synopsis    15583 non-null  object 
 4   Type        16368 non-null  object 
 5   Producer    7635 non-null   object 
 6   Studio      7919 non-null   object 
 7   Rating      14425 non-null  float64
 8   ScoredBy    13227 non-null  float64
 9   Popularity  16368 non-null  float64
 10  Members     17002 non-null  float64
 11  Episodes    14085 non-null  float64
 12  Source      15075 non-null  object 
 13  Aired       16368 non-null  object 
 14  Link        16368 non-null  object 
dtypes: float64(5), int64(1), object(9)
memory usage: 1.9+ MB


In [4]:
data.isna().sum()

Anime_id         0
Title            0
Genre         2012
Synopsis      1419
Type           634
Producer      9367
Studio        9083
Rating        2577
ScoredBy      3775
Popularity     634
Members          0
Episodes      2917
Source        1927
Aired          634
Link           634
dtype: int64

In [5]:
data.shape

(17002, 15)

In [6]:
data['Anime_id'].value_counts().unique()

array([2, 1], dtype=int64)

In [7]:
data.describe()

Anime_id        Rating      ScoredBy    Popularity       Members  \
count  17002.000000  14425.000000  1.322700e+04  16368.000000  1.700200e+04   
mean   20446.579638      6.287867  1.139084e+04   8131.919599  2.038130e+04   
std    14342.513259      1.141401  4.328434e+04   4714.683351  7.121404e+04   
min        1.000000      1.000000  1.000000e+00      1.000000  0.000000e+00   
25%     5581.500000      5.620000  4.300000e+01   4042.500000  1.450000e+02   
50%    21334.000000      6.410000  4.780000e+02   8115.000000  1.113000e+03   
75%    34789.250000      7.090000  3.831000e+03  12208.250000  7.855750e+03   
max    40960.000000     10.000000  1.006242e+06  16338.000000  1.451708e+06   

           Episodes  
count  14085.000000  
mean      11.482712  
std       44.089040  
min        1.000000  
25%        1.000000  
50%        1.000000  
75%       12.000000  
max     1818.000000

In [8]:
data['Anime_id'].duplicated().sum()

152

In [9]:
data['Type'].unique()

array(['TV', 'Movie', 'OVA', 'Special', 'ONA', 'Music', 'Unknown', nan],
      dtype=object)

We will create a function that will be used to clean the anime titles using REGEX.

In [10]:
# Function for cleaning the anime titles with REGEX
import re

def clean_title(title):
    title = re.sub("[^a-zA-Z0-9 ]", "", title)
    return title

In [11]:
# applying the function on the title column
data["Title"] = data["Title"].apply(clean_title)
data

Anime_id                                          Title  \
0             1                                   Cowboy Bebop   
1             5                 Cowboy Bebop Tengoku no Tobira   
2             6                                         Trigun   
3             7                             Witch Hunter Robin   
4             8                                 Bouken Ou Beet   
...         ...                                            ...   
16997     27499  Sore Ike Anpanman Anpanman to Kaizoku Lobster   
16998     12091                                       X Bomber   
16999     29133                                 X Bomber Pilot   
17000     34485                                      Gankochan   
17001     32713                            Original CVP Momoko   

                                                   Genre  \
0      ['Action', 'Adventure', 'Comedy', 'Drama', 'Sc...   
1      ['Action', 'Space', 'Drama', 'Mystery', 'Sci-Fi']   
2      ['Action', 'Sci-Fi', 'Adventure', 'Comedy', 'D...   
3      ['Action', 'Magic', 'Police', 'Supernatural', ...   
4      ['Adventure', 'Fantasy', 'Shounen', 'Supernatu...   
...                                                  ...   
16997                                                NaN   
16998                                                NaN   
16999                                                NaN   
17000                                                NaN   
17001                                                NaN   

                                                Synopsis     Type  \
0      In the year 2071, humanity has colonized sever...       TV   
1      Another day, another bounty—such is the life o...    Movie   
2      Vash the Stampede is the man with a $$60,000,0...       TV   
3      Witches are individuals with special powers li...       TV   
4      It is the dark century and the people are suff...       TV   
...                                                  ...      ...   
16997                                                NaN  Special   
16998                                                NaN       TV   
16999                                                NaN  Special   
17000                                                NaN      NaN   
17001                                                NaN      OVA   

                           Producer              Studio  Rating  ScoredBy  \
0                 ['Bandai Visual']         ['Sunrise']    8.81  363889.0   
1      ['Sunrise', 'Bandai Visual']           ['Bones']    8.41  111187.0   
2          ['Victor Entertainment']        ['Madhouse']    8.31  197451.0   
3                 ['Bandai Visual']         ['Sunrise']    7.34   31875.0   
4                               NaN  ['Toei Animation']    7.04    4757.0   
...                             ...                 ...     ...       ...   
16997                           NaN                 NaN    5.50       NaN   
16998                           NaN                 NaN    5.94       NaN   
16999                           NaN                 NaN    4.28       NaN   
17000                           NaN                 NaN     NaN       NaN   
17001                           NaN                 NaN    4.00       NaN   

       Popularity   Members  Episodes    Source                         Aired  \
0            39.0  704490.0      26.0  Original   Apr 3, 1998 to Apr 24, 1999   
1           475.0  179899.0       1.0  Original                   Sep 1, 2001   
2           158.0  372709.0      26.0     Manga   Apr 1, 1998 to Sep 30, 1998   
3          1278.0   74889.0      26.0  Original   Jul 2, 2002 to Dec 24, 2002   
4          3968.0   11247.0      52.0     Manga  Sep 30, 2004 to Sep 29, 2005   
...           ...       ...       ...       ...                           ...   
16997         NaN      70.0       NaN       NaN                           NaN   
16998         NaN     199.0       NaN       NaN                           NaN   
16999     

We will then use a Term frequency - inverse document frequency to create a search engine so that it is easy to find an anime title and its anime id. The tfdif will enable the computer to find the title that is most similar to the title that we will enter.

In [12]:
# vector matrix for the anime titles

from sklearn.feature_extraction.text import TfidfVectorizer

# finding groups of 2 words that are consecutive
vectorizer = TfidfVectorizer(ngram_range=(1,2))

tfidf = vectorizer.fit_transform(data["Title"])

We will then create a function that will use cosine similarity to compute the similarity between a term that we will enter in our search box and all the anime titles in our dataset.

In [13]:
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

# creating a search function 
def search(title):
    """
    A function that takes in a search term of the title that we are 
    looking for then cleans it and uses the vectorizer
    to turn it into a sparse matrix and then uses cosine similarity to find the
    most similar titles.
    """
    # cleaning the title entered
    title = clean_title(title)
    query_vec = vectorizer.transform([title])
    similarity = cosine_similarity(query_vec, tfidf).flatten()

    # find the 5 most similar titles to the searcch term
    indices = np.argpartition(similarity, -5)[-5:]

    # reverse results so that most similar result is at the top
    results = data.iloc[indices].iloc[::-1]
    
    return results

We will then build an interactive search box using our function above. The search box will enable one to type in the name of an anime and see the results.

In [14]:
import ipywidgets as widgets
from IPython.display import display

movie_input = widgets.Text(
    # setting default value as Trigun
    value='Trigun',
    description='Movie Title:',
    disabled=False
)
movie_list = widgets.Output()

def on_type(data):
    with movie_list:
        movie_list.clear_output()
        title = data["new"]

        # display output when length of title is more than 5
        if len(title) > 5:
            display(search(title))

movie_input.observe(on_type, names='value')


display(movie_input, movie_list)

Text(value='Trigun', description='Movie Title:')

Output()

We will hardcode an anime id for the 'Cowoy Bebop' anime to find the users who watched and also liked same anime and gave it a rating greater than 6.

In [15]:
# hard coding an anime id to find the users who liked the same movie
movie_id = 1

#def find_similar_movies(movie_id):
anime = data[data["Anime_id"] == movie_id]

We  will now use the ratings dataset to find the users who liked the same movie.

In [16]:
ratings = pd.read_csv('rating.csv')
ratings.head()

user_id  anime_id  rating
0        1        20      -1
1        1        24      -1
2        1        79      -1
3        1       226      -1
4        1       241      -1

In [17]:
# finding similar users
similar_users = ratings[(ratings["anime_id"] == movie_id) & (ratings["rating"] > 6)]["user_id"].unique()
similar_users

array([   19,    21,    23, ..., 73507, 73513, 73515], dtype=int64)

We will then create a put it into a dataframe to return the records of the similar users.

In [18]:

similar_user_recs = ratings[(ratings["user_id"].isin(similar_users)) & (ratings["rating"] > 6)]
similar_user_recs

user_id  anime_id  rating
2240          19         1      10
2241          19         5       9
2242          19       121       8
2243          19       199       9
2244          19       578      10
...          ...       ...     ...
7813730    73515     13659       8
7813731    73515     14345       7
7813732    73515     16512       7
7813733    73515     17187       9
7813734    73515     22145      10

[1809323 rows x 3 columns]

In [19]:
# getting the anime ids from the dataframe

similar_user_recs = ratings[(ratings["user_id"].isin(similar_users)) & (ratings["rating"] > 6)]["anime_id"]
similar_user_recs

2240           1
2241           5
2242         121
2243         199
2244         578
           ...  
7813730    13659
7813731    14345
7813732    16512
7813733    17187
7813734    22145
Name: anime_id, Length: 1809323, dtype: int64

We will then narrow down the dataframe so that we only get 10% or more of users who are similar and also liked the same movie

In [20]:
# get value counts for the movie ids
similar_user_recs.value_counts()

1        12940
1535      8315
1575      7235
121       7011
2001      6785
         ...  
30840        1
25173        1
11869        1
7771         1
30178        1
Name: anime_id, Length: 8239, dtype: int64

In [21]:
# Converting to percentage
similar_user_recs = similar_user_recs.value_counts() / len(similar_users)

# Taking records where only greater than 10% liked
similar_user_recs = similar_user_recs[similar_user_recs > .10]
similar_user_recs

1        1.000000
1535     0.642581
1575     0.559119
121      0.541808
2001     0.524343
           ...   
3002     0.101391
486      0.101159
239      0.100773
357      0.100696
15451    0.100077
Name: anime_id, Length: 326, dtype: float64

We now have a set of 326 animes where more than 10% of users liked the anime title. We will now find users who liked animes that are similar to the 'Cowboy' anime. This will be done by finding the users who have rated the animes in our 'similar_user_recs'.

In [22]:
# finding how much all users in the dataset like the animes
all_users = ratings[(ratings["anime_id"].isin(similar_user_recs.index)) & (ratings["rating"] > 6)]
all_users

user_id  anime_id  rating
47             1      8074      10
81             1     11617      10
83             1     11757      10
101            1     15451      10
153            2     11771      10
...          ...       ...     ...
7813723    73515     12413       9
7813730    73515     13659       8
7813731    73515     14345       7
7813735    73516       790       9
7813736    73516      8074       9

[2337673 rows x 3 columns]

We will then find the percentage of all users who recommended the animes that are in ``similar_user_recs``.

In [23]:
# calculating percentage
all_user_recs = all_users["anime_id"].value_counts() / len(all_users["user_id"].unique())
all_user_recs

1535     0.481432
16498    0.353069
1575     0.343044
11757    0.329717
6547     0.322566
           ...   
166      0.034704
486      0.030753
239      0.030679
617      0.030473
329      0.028866
Name: anime_id, Length: 326, dtype: float64

We will then compare the percentages in ```similar_user_recs``` and ```all_users_recs``` by joining the two pandas series so as to see how users similar to us liked the anime and how much an average person liked the same animes.

The goal is to find the animes that have a big difference.

In [24]:
# joining the two pandas series
rec_percentages = pd.concat([similar_user_recs, all_user_recs], axis=1)
rec_percentages.columns = ["similar", "all"]
rec_percentages

similar       all
1      1.000000  0.190768
5      0.382767  0.081025
6      0.431453  0.130427
16     0.113524  0.044685
19     0.165533  0.057894
...         ...       ...
30276  0.263292  0.162522
30503  0.105564  0.079079
31043  0.168161  0.112839
31240  0.111283  0.073683
31964  0.109892  0.070322

[326 rows x 2 columns]

We will then calculate a score by dividing one percentage by the other.

In [25]:
# finding the ratio between the percentages
rec_percentages["score"] = rec_percentages["similar"] / rec_percentages["all"]

# sorting the recommendations based on the scores
rec_percentages = rec_percentages.sort_values("score", ascending=False)
rec_percentages

similar       all     score
1      1.000000  0.190768  5.241963
5      0.382767  0.081025  4.724062
400    0.142040  0.035721  3.976363
329    0.106105  0.028866  3.675799
801    0.184158  0.050125  3.673999
...         ...       ...       ...
11757  0.339335  0.329717  1.029173
15451  0.100077  0.098170  1.019424
11617  0.147991  0.147484  1.003435
7054   0.165688  0.174109  0.951632
3457   0.129212  0.138639  0.932004

[326 rows x 3 columns]

From the above dataframe, the higher the score, the better the recommendation.

We will then get the top 10 recommendations and merge that with our ```anime``` dataset so as to get the title of the movies.

In [26]:
# getting top 10 recommendations
rec_percentages.head(10).merge(data, left_index=True, right_on="Anime_id")

similar       all     score  Anime_id  \
0      1.000000  0.190768  5.241963         1   
1      0.382767  0.081025  4.724062         5   
15087  0.142040  0.035721  3.976363       400   
301    0.106105  0.028866  3.675799       329   
653    0.184158  0.050125  3.673999       801   
3226   0.125502  0.035426  3.542633      4106   
406    0.263756  0.074877  3.522508       467   
527    0.106337  0.030473  3.489570       617   
15078  0.150850  0.043741  3.448706       468   
244    0.128130  0.037402  3.425768       267   

                                              Title  \
0                                      Cowboy Bebop   
1                    Cowboy Bebop Tengoku no Tobira   
15087                     Seihou Bukyou Outlaw Star   
301                                        Planetes   
653    Koukaku Kidoutai Stand Alone Complex 2nd GIG   
3226                         Trigun Badlands Rumble   
406            Koukaku Kidoutai Stand Alone Complex   
527                               Juubee Ninpuuchou   
15078                  Koukaku Kidoutai 2 Innocence   
244                                        Gungrave   

                                                   Genre  \
0      ['Action', 'Adventure', 'Comedy', 'Drama', 'Sc...   
1      ['Action', 'Space', 'Drama', 'Mystery', 'Sci-Fi']   
15087                                                NaN   
301    ['Drama', 'Romance', 'Sci-Fi', 'Seinen', 'Space']   
653    ['Action', 'Military', 'Sci-Fi', 'Mystery', 'P...   
3226   ['Action', 'Sci-Fi', 'Adventure', 'Comedy', 'D...   
406    ['Action', 'Military', 'Sci-Fi', 'Police', 'Me...   
527    ['Adventure', 'Historical', 'Horror', 'Superna...   
15078                                                NaN   
244    ['Action', 'Drama', 'Sci-Fi', 'Seinen', 'Super...   

                                                Synopsis   Type  \
0      In the year 2071, humanity has colonized sever...     TV   
1      Another day, another bounty—such is the life o...  Movie   
15087  Gene Starwind has always dreamed of piloting h...     TV   
301    In 2075, space travel is no longer just a drea...     TV   
653    Following the closure of the "Laughing Man" ca...     TV   
3226    20 years after meddling into the bank heist o...  Movie   
406    In the not so distant future, mankind has adva...     TV   
527    Jubei Kibagami wanders feudal Japan as an itin...  Movie   
15078  With Major Motoko Kusanagi missing, Section 9'...  Movie   
244    Brandon Heat and Harry MacDowel, two friends s...     TV   

                                                Producer              Studio  \
0                                      ['Bandai Visual']         ['Sunrise']   
1                           ['Sunrise', 'Bandai Visual']           ['Bones']   
15087                                                NaN                 NaN   
301                             ['Bandai Visual', 'NHK']         ['Sunrise']   
653    ['Bandai Visual', 'Dentsu', 'Victor Entertainm...  ['Production I.G']   
3226   ['flying DOG', 'Sunny Side Up', 'Studio Jack',...        ['Madhouse']   
406    ['Bandai Visual', 'Dentsu', 'Victor Entertainm...  ['Production I.G']   
527                                            ['Movic']        ['Madhouse']   
15078                                                NaN                 NaN   
244                   ['Tohokushinsha Film Corporation']        ['Madhouse']   

       Rating  ScoredBy  Popularity   Members  Episodes    Source  \
0        8.81  363889.0        39.0  704490.0      26.0  Original   
1        8.41  111187.0       475.0  179899.0       1.0  Original   
15087    7.98       NaN      1016.0  109861.0       NaN       NaN   
301      8.36   46131.0       656.0  138203.0      26.0     Manga   
653      8.55   72115.0       626.0  149102.0      26.0     Manga   
3226     8.05   46942.0      1117.0   84011.0       1.0     Manga   
406      8.46  110151.0       331.0  251675.0      26.0     Manga   
527      7.66   40341.0      1

We will then put all the code above into one function.

In [27]:
def find_similar_movies(movie_id):
    
    # finding recommendations for users that are similar 
    similar_users = ratings[(ratings["anime_id"] == movie_id) & (ratings["rating"] > 6)]["user_id"].unique()
    similar_user_recs = ratings[(ratings["user_id"].isin(similar_users)) & (ratings["rating"] > 6)]["anime_id"]

    # narrowing them down to 10%
    similar_user_recs = similar_user_recs.value_counts() / len(similar_users)
    similar_user_recs = similar_user_recs[similar_user_recs > .10]

    # finding how common the recommendations are among all users
    all_users = ratings[(ratings["anime_id"].isin(similar_user_recs.index)) & (ratings["rating"] > 4)]
    all_user_recs = all_users["anime_id"].value_counts() / len(all_users["user_id"].unique())

    # concatenating all_user_recs and similar_user_recs
    rec_percentages = pd.concat([similar_user_recs, all_user_recs], axis=1)
    rec_percentages.columns = ["similar", "all"]
    
    # calculating the score
    rec_percentages["score"] = rec_percentages["similar"] / rec_percentages["all"]
    # sorting values based on score
    rec_percentages = rec_percentages.sort_values("score", ascending=False)

    # return the top 10 recommendations
    return rec_percentages.head(10).merge(data, left_index=True, right_on="Anime_id")[["score", "Title", "Genre"]]

We will now create an interactive recommendation widget where one gets to type in an anime title and get recommendations based on the anime title

In [30]:
# creating an input widget
movie_name_input = widgets.Text(
    value='Trigun',
    description='Movie Title:',
    disabled=False
)

# creating an output widget
recommendation_list = widgets.Output()

# creating an ontype function
def on_type(data):
    with recommendation_list:
        recommendation_list.clear_output()

        # grab title from input widget
        title = data["new"]

        # display output when length of title is more than 5
        if len(title) > 5:
            # search title using the search function
            results = search(title)
            movie_id = results.iloc[0]["Anime_id"]
            display(find_similar_movies(movie_id))

movie_name_input.observe(on_type, names='value')

display(movie_name_input, recommendation_list)

Text(value='Trigun', description='Movie Title:')

Output()

In [29]:
import pickle
with open ('similar_movies.pkl','wb') as f:
    pickle.dump(find_similar_movies,f)